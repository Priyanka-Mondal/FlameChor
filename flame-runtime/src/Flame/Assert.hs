{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE PostfixOperators #-}
{-# LANGUAGE FlexibleContexts #-}
{-# OPTIONS_GHC -fplugin Flame.Solver -fobject-code #-}
module Flame.Assert
where
  
import Flame.Principals
assertEq :: (l === l') => SPrin l -> SPrin l' -> ()
assertEq l l' = ()

assertActsFor :: (p ≽ q) => SPrin p -> SPrin q -> ()
assertActsFor p q = ()

assertFlowsTo :: (l ⊑ l') => SPrin l -> SPrin l' -> ()
assertFlowsTo l l' = ()
