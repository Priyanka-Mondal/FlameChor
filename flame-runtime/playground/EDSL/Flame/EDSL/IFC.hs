{-# LANGUAGE GADTs, PolyKinds, KindSignatures, MultiParamTypeClasses,
    DataKinds, RankNTypes, FlexibleInstances, FlexibleContexts, TypeFamilies #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE TypeOperators, PostfixOperators #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE UndecidableSuperClasses #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE TemplateHaskell #-}
{-# OPTIONS_GHC -fplugin Flame.Solver #-}

module Flame.EDSL.IFC where

import Data.Proxy (Proxy(..))
import Data.Constraint
import Data.Constraint.Unsafe
import Data.Reflection

import Flame.Principals
-- import Flame.TCB.Assume
import Data.Int
import Data.Functor.Identity
import Text.PrettyPrint.Mainland

{- EDSL imports -}
import Control.Monad.Operational.Higher
import qualified Control.Monad.State as CMS
import Control.Monad.State.Class

import Language.C.Quote.C
import qualified Language.C.Syntax as C

import Language.C.Monad hiding (inModule)
import qualified Control.Monad.Reader as R
import Language.Embedded.Expression
--import Language.Embedded.Imperative hiding (cedecl, RefCMD, Ref)
import Language.Embedded.Backend.C
import Language.Embedded.CExp
import Language.Embedded.Traversal
import Language.Embedded.Imperative.CMD (RefCMD (GetRef), mkParam, mapFunArg, mapFunArgM)

import Flame.Principals

data Label (l::KPrin) a where
  Label   :: a -> Label l a
  Unlabel :: (l ⊑ l') => Label l a -> (a -> Label l' b) -> Label l' b

runLabel :: Label l a -> a
runLabel (Label v) = v
runLabel (Unlabel lv f) = runLabel (f $ runLabel lv)

newtype LABProgram exp instr pred pc l a = LAB { program :: Program instr (Param2 exp pred) a }

lift :: Label l a -> LABProgram exp instr pred pc l a
lift = LAB . return . runLabel

apply :: (pc ⊑ pc', pc ⊑ pc'') => LABProgram exp instr pred pc l a
      -> (Label l a -> LABProgram exp instr pred pc' l' b)
      -> LABProgram exp instr pred pc'' l' b
apply (LAB prog) f = LAB $ do
                             lv <- prog
                             case f (Label lv) of
                               LAB p -> p

bind ::  (l ⊑ l', l ⊑ pc') => Label l a
     -> (a -> LABProgram exp instr pred pc' l' b)
     -> LABProgram exp instr pred pc l' b
bind lv f = LAB $ case f $ runLabel lv of
                    LAB p -> p

class HasBackend exp1 exp2 instr pred where
  translateExp :: exp1 a -> Program instr (Param2 exp2 pred) (exp2 a)

reexpressLAB :: forall instr1 instr2 exp1 exp2 pred pc l a b .
              (Reexpressible instr1 instr2 ()
              , HasBackend exp1 exp2 instr2 pred
              )
             => (forall b . exp1 b -> Program instr2 (Param2 exp2 pred) (exp2 b))
             -> LABProgram exp1 instr1 pred pc l a -> LABProgram exp2 instr2 pred pc l a
reexpressLAB f (LAB p) = LAB $ reexpress @instr1 @instr2 @_ @exp1 @exp2 f p

wrapProc :: MonadC m => String -> m a -> m () 
wrapProc s prog = do
    (_,uvs,params,items) <- inNewFunction $ prog >> addStm [cstm| return 0; |]
    setUsedVars s uvs
    addGlobal [cedecl| int $id:s($params:params){ $items:items }|]

compileLAB :: forall instr exp pred pc l.
              ( HasBackend exp CExp instr pred 
              , Reexpressible instr instr ()
              , Interp instr CGen (Param2 CExp pred)
              , FreeExp exp, RefCMD :<: instr
              )
           => String -> LABProgram exp instr pred pc l () -> [(String,Doc)]
compileLAB s p = prettyCGen . wrapProc s $ (interpret . program) (reexpressLAB @instr @instr @exp @CExp translateExp p)

relabel :: (l ⊑ l') => Label l a -> Label l' a
relabel a = Unlabel a Label 

protect :: (pc ⊑ l) => a -> LABProgram exp instr pred pc l a
protect = lift . Label

use :: forall exp instr pred l l' pc pc' pc'' a b.
       (l ⊑ l', pc ⊑ pc', l ⊑ pc', pc ⊑ pc'')
    => LABProgram exp instr pred pc l a -> (a -> LABProgram exp instr pred pc' l' b) -> LABProgram exp instr pred pc'' l' b
use x f = apply x $ \x' -> (bind x' f :: LABProgram exp instr pred pc' l' b)

(>>>=) :: forall exp instr pred l l' pc pc' pc'' a b.
       (l ⊑ l', pc ⊑ pc', l ⊑ pc', pc ⊑ pc'')
    => LABProgram exp instr pred pc l a -> (a -> LABProgram exp instr pred pc' l' b) -> LABProgram exp instr pred pc'' l' b
x >>>= f = use x f
 
reprotect :: forall exp instr pred l l' pc pc' a.
             (l ⊑ l', pc ⊑ pc', (pc ⊔ l) ⊑ l')
          => LABProgram exp instr pred pc l a -> LABProgram exp instr pred pc' l' a 
reprotect x = use x (protect :: a -> LABProgram exp instr pred (pc ⊔ l) l' a)

ifmap :: forall exp instr pred l l' pc pc' a b.
         (l ⊑ l', pc ⊑ pc', l ⊑ pc', pc' ⊑ l')
      => (a -> b) -> LABProgram exp instr pred pc l a -> LABProgram exp instr pred pc' l' b
ifmap f x = use x (\x' -> protect (f x') :: LABProgram exp instr pred (pc ⊔ l) l' b)

ijoin :: forall exp instr pred l l' pc pc' a. (l ⊑ l',  pc ⊑ pc', l ⊑ pc')
      => LABProgram exp instr pred pc l (LABProgram exp instr pred pc' l' a) -> LABProgram exp instr pred pc' l' a
ijoin x = use x id
