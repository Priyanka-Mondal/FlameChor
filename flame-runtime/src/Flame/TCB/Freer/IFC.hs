{-# LANGUAGE GADTs, PolyKinds, KindSignatures, MultiParamTypeClasses,
    DataKinds, RankNTypes, FlexibleInstances, FlexibleContexts, TypeFamilies #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE TypeOperators, PostfixOperators #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE UndecidableSuperClasses #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE BangPatterns #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE PackageImports #-}

{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE UndecidableInstances #-}

{-# OPTIONS_GHC -fplugin Flame.Solver -fobject-code #-}

module Flame.TCB.Freer.IFC where

import Flame.Principals
import "HasChor" Control.Monad.Freer
    ( interpFreer, toFreer, Freer )
import "HasChor" Control.Monad.Freer
    ( interpFreer, toFreer, Freer )
import Data.Type.Bool
import Data.Proxy (Proxy(..))
import Data.Kind (Type)
import Data.Functor
import Data.Constraint
import Data.Constraint.Unsafe
import Data.Reflection
-- import Data.Singletons (Apply, TyFun)
import GHC.TypeLits (TypeError, ErrorMessage(..))
import System.IO
import Control.Monad ((>=>))
import Control.Monad.Identity


-- instance (Member (Labeled pc) r) => LabeledMember pc r where
data (l::KPrin) ! a  where
  Seal :: { unseal :: a }  -> l!a
  deriving (Show, Read) -- these instances should be explicitly
                        -- most likely, they should encrypt and decrypt

type Clearance pc = forall a l. (l ⊑ pc) => l!a -> a

-- type Use pc = forall pc a l. (l⊑pc) => (l ! a) -> a
data LabeledSig m (pc::KPrin) a where
    Restrict :: (Monad m, pc ⊑ l) => !(SPrin pc) -> (Clearance pc -> m a) -> LabeledSig m pc (l!a)
    Protect  :: (Monad m, pc ⊑ l) => a -> LabeledSig m pc (l!a)
    Use      :: (Monad m, l' ⊑ l, l' ⊑ pc') =>
      l'!b -> (b -> Labeled m pc' (l!a)) -> LabeledSig m pc (l!a)

type Labeled m pc = Freer (LabeledSig m pc)

restrict :: SPrin pc -> (Monad m, pc ⊑ l) => (Clearance pc -> m a) -> Labeled m pc (l!a)
restrict pc ma = toFreer (Restrict pc ma)

protect :: (Monad m, pc ⊑ l) => a -> Labeled m pc (l!a)
protect a = toFreer (Protect a)

use :: (Monad m, l' ⊑ l, l' ⊑ pc') => l'!b -> (b -> Labeled m pc' (l!a)) -> Labeled m pc (l!a)
use b k = toFreer (Use b k)

label :: forall l a. a -> l!a
label = runIdentity . runLabeled . protect

relabel :: forall l l' a. l ⊑ l' => l!a -> l'!a
relabel = runIdentity . runLabeled . \x -> use x protect

--- Is this secure? 
-- runLabeled $ join' $ restrict high $ \un -> 
--    if (un secret) then (runLabeled lowT) else (runLabeled lowF)
      
relabel' :: forall l l' pc pc' m a. (Monad m, l ⊑ l', pc ⊑ l') => 
              SPrin pc -> Labeled m pc' (l!a) -> Labeled m pc (l'!a)
relabel' pc e = join' $ restrict pc $ \_ -> do 
  x <- runLabeled e
  return $ relabel x
  
join :: forall l l' l'' a. (l ⊑ l'', l' ⊑ l'') => l!(l'!a) -> l''!a
join = runIdentity . runLabeled . \x -> use x (`use` protect)

join' :: forall l l' l'' pc m a. (Monad m, l ⊑ l'', l' ⊑ l'') => Labeled m pc (l!(l'!a)) -> Labeled m pc (l''!a)
join' lx = lx >>= (\x -> use x (`use` protect))

bind :: forall l l' a b. l ⊑ l' => l!a -> (a -> l'!b) -> l'!b
bind la k = runIdentity . runLabeled $ use la (join' @_ @l' @l' @l' . protect . k)

fmap :: (Monad m, l ⊑ l', pc ⊑ pc', l ⊑ pc', pc' ⊑ l') =>
    (a -> b) -> l!a -> l'!b
fmap f = runIdentity . runLabeled . (`use` (protect . f))

runLabeled :: forall pc m a. Monad m => Labeled m pc a -> m a
runLabeled = interpFreer handler
  where
    handler :: forall pc m a. Monad m => LabeledSig m pc a -> m a
    handler (Restrict pc ma) = ma unseal <&> Seal
    handler (Protect a) = pure (Seal a)
    handler (Use (Seal b) k)  = runLabeled $ k b

-- chooseSecret :: (l ⊑ pc, l' ⊑ pc) => 
--     SPrin pc -> SPrin l -> SPrin l' -> l!Bool -> l'!Int -> l'!Int -> Labeled IO pc (pc!())
-- chooseSecret pc l l' lb n1 n2 = use lb
--      (\b ->  if b then
--                  relabel' $ s_putStrLn n1 
--              else 
--                  relabel' $ s_putStrLn n2)
--    where 
--      s_putStrLn la = restrict pc (\un -> putStrLn (show (un la)))
--         
-- 
-- type Alice = KName "Alice"
-- alice :: SPrin Alice
-- alice = SName (Proxy :: Proxy "Alice")
-- type Bob = KName "Bob"
-- bob :: SPrin Bob
-- bob = SName (Proxy :: Proxy "Bob")
-- 
-- main :: IO (Alice!())
-- main = runLabeled $ chooseSecret alice alice bob (Seal True :: Alice!Bool) (Seal 1 :: Bob!Int) (Seal 2 :: Bob!Int)