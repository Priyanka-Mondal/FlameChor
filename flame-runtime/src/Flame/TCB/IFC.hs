{-# LANGUAGE GADTs, PolyKinds, KindSignatures, MultiParamTypeClasses,
    DataKinds, RankNTypes, FlexibleInstances, FlexibleContexts, TypeFamilies #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE TypeOperators, PostfixOperators #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE InstanceSigs #-}
{-# OPTIONS_GHC -fplugin Flame.Solver -fobject-code #-}

module Flame.TCB.IFC where

import Data.Proxy (Proxy(..))
import Data.Constraint
import Data.Constraint.Unsafe
import Data.Reflection

import Flame.Principals
import Flame.TCB.Assume

{- An indexed monad for information flow on pure computation -}
class Labeled (n :: KPrin -> * -> *) where
  label     :: a -> n l a
  unlabel :: (l ⊑ l') => n l a -> (a -> n l' b) -> n l' b
  unsafeUnlabel :: n l a -> a
  
  relabel :: (l ⊑ l') => n l a -> n l' a
  relabel a = unlabel a label

{- Information flow control based on FLAM acts-for constraints -}
class (Monad e, Labeled n) => IFC (m :: (* -> *) -> (KPrin -> * -> *) -> KPrin -> KPrin -> * -> *) e n where
  lift   :: (pc ⊑ l) => n l a -> m e n pc l a

  apply  :: (pc ⊑ pc', pc ⊑ pc'') => m e n pc l a -> (n l a -> m e n pc' l' b) -> m e n pc'' l' b

  bind  :: (l ⊑ l', l ⊑ pc) => n l a -> (a -> m e n pc l' b) -> m e n pc' l' b

  unsafeProtect :: e (n l a) -> m e n pc l a

  runIFC :: m e n pc l a -> e (n l a)

  protect :: (pc ⊑ l) => a -> m e n pc l a
  protect = lift . label

  use :: forall l l' pc pc' pc'' a b. (l ⊑ l', pc ⊑ pc', l ⊑ pc', pc ⊑ pc'') =>
         m e n pc l a -> (a -> m e n pc' l' b) -> m e n pc'' l' b
  use x f = apply x $ \x' -> (bind x' f :: m e n pc' l' b)
 
  reprotect :: forall l l' pc pc' a. (l ⊑ l', pc ⊑ pc', (pc ⊔ l) ⊑ l') => m e n pc l a -> m e n pc' l' a 
  reprotect x = use x (protect :: a -> m e n (pc ⊔ l) l' a)

  ffmap :: forall l l' pc pc' a b. (l ⊑ l', pc ⊑ pc', l ⊑ pc', pc' ⊑ l') => (a -> b) -> m e n pc l a -> m e n pc' l' b
  ffmap f x = use x (\x' -> protect (f x') :: m e n (pc ⊔ l) l' b)

  fjoin  :: forall l l' pc pc' a. (l ⊑ l',  pc ⊑ pc', l ⊑ pc') => m e n pc l (m e n pc' l' a) -> m e n pc' l' a
  fjoin x = use x id

{- Flow-limited authorization for IFC types -}
class IFC m e n => FLA m e n where
  weak_assume :: (pc ≽ ((I q) ∧ (∇) q), (∇) p ≽ (∇) q) =>
              (p :≽ q) -> ((p ≽ q) => m e n pc l a) -> m e n pc l a
  weak_assume = unsafeAssume
  assume :: (pc ≽ ((I q) ∧ (∇) q)) =>
              (p :≽ q) -> ((p ≽ q, (∇) p ≽ (∇) q) => m e n pc l a) -> m e n pc l a
  assume (Del p q) f = unsafeAssume ((*∇) p ≽ (*∇) q) $ unsafeAssume (p ≽ q) f


{- Nonmalleable information flow control -}
class IFC m e n => NMIF m e n where
  declassify :: (((C pc) ⊓ (C l')) ⊑ (C l)
                 , (C pc) ⊑ (C l')
                 , (C l') ⊑ ((C l) ⊔ (Δ (I l' ⊔ I pc)))
                 , (I l') === (I l)) =>
             m e n pc l' a -> m e n pc l a 
  declassify x = unsafeProtect $ do
    x' <- runIFC x 
    return $ label (unsafeUnlabel x')
  endorse    :: ( ((I pc) ⊓ (I l')) ⊑ (I l)
                , (I pc) ⊑ (I l')
                , (I l') ⊑ ((I l) ⊔ ((∇) (C l' ⊔ C pc)))
                , (C l') === (C l)
                ) =>
             m e n pc l' a -> m e n pc l a 
  endorse x = unsafeProtect $ do
    x' <- runIFC x 
    return $ label (unsafeUnlabel x')

{- A type for pure labeled computations -}
-- both the constructor and destructor must be private:
--   since either will unlabel a value
--   (the constructor enables pattern matching)
data Lbl (l::KPrin) a where
  UnsafeMkLbl :: { unsafeRunLbl :: a } -> Lbl l a

lbl_label :: a -> Lbl l a
lbl_label = UnsafeMkLbl

lbl_unlabel :: (l ⊑ l') => Lbl l a -> (a -> Lbl l' b) -> Lbl l' b    
lbl_unlabel x f = f (unsafeRunLbl x)

{- A Labeled instance for Lbl -}
instance Labeled Lbl where
  label = lbl_label
  unlabel = lbl_unlabel
  unsafeUnlabel = unsafeRunLbl

{- A transformer for effectful labeled computations -}
data FLACT e (n :: KPrin -> * -> *) (pc :: KPrin) (l :: KPrin) a where
  UnsafeFLAC :: { runFLAC :: e (n l a) } -> FLACT e n pc l a

{- FLAC computations -}
type FLAC e pc l a = FLACT e Lbl pc l a

flac_lift :: Monad e => Lbl l a -> FLAC e pc l a
flac_lift  x = UnsafeFLAC $ Prelude.return x

flac_bind :: (Monad e, l ⊑ l', l ⊑ pc') => Lbl l a -> (a -> FLAC e pc' l' b) -> FLAC e pc l' b
flac_bind x f = UnsafeFLAC $ runFLAC $ f $ unsafeRunLbl x

flac_apply :: (Monad e, pc ⊑ pc', pc ⊑ pc'') => FLAC e pc l a -> (Lbl l a -> FLAC e pc' l' b) -> FLAC e pc'' l' b
flac_apply x f = UnsafeFLAC $ do
                   a <- runFLAC x
                   runFLAC $ f a

instance Monad e => IFC FLACT e Lbl where
  lift  = flac_lift 
  apply = flac_apply
  bind  = flac_bind
  unsafeProtect = UnsafeFLAC
  runIFC  = runFLAC

instance Monad e => FLA FLACT e Lbl where {}

{- A transformer for effectful labeled computations -}
data NMT e (n :: KPrin -> * -> *) (pc :: KPrin) (l :: KPrin) a where
  UnsafeNM :: { runNM :: e (n l a) } -> NMT e n pc l a

type NM e pc l a = NMT e Lbl pc l a

nmif_lift :: Monad e => Lbl l a -> NM e pc l a
nmif_lift  x = UnsafeNM $ Prelude.return x

nmif_bind :: (Monad e, l ⊑ l', l ⊑ pc') => Lbl l a -> (a -> NM e pc' l' b) -> NM e pc l' b
nmif_bind x f = UnsafeNM $ runNM $ f $ unsafeRunLbl x

nmif_apply :: (Monad e, pc ⊑ pc', pc ⊑ pc'') => NM e pc l a -> (Lbl l a -> NM e pc' l' b) -> NM e pc'' l' b
nmif_apply x f = UnsafeNM $ do
                   a <- runNM x
                   runNM $ f a

instance Monad e => IFC NMT e Lbl where
  lift    = nmif_lift 
  bind   = nmif_bind
  apply   = nmif_apply
  unsafeProtect = UnsafeNM
  runIFC  = runNM

instance Monad e => NMIF NMT e Lbl where {}