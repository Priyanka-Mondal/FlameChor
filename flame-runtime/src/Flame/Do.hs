{-# LANGUAGE TemplateHaskell #-}
{-# OPTIONS_GHC -fplugin Flame.Solver -fobject-code #-}

----------------------------------------------------------------------------
---- Based on Language.Haskell.IndexedDo by Fumiaki Kinoshita
----------------------------------------------------------------------------
module Flame.Do (flame) where

import Language.Haskell.Meta
import Language.Haskell.TH
import Language.Haskell.TH.Quote
import qualified Language.Haskell.Exts.Syntax as Hs
import qualified Language.Haskell.Exts.SrcLoc as Hs
import Language.Haskell.Exts.Extension as Ext
import Language.Haskell.Exts.Parser (ParseMode(..), ParseResult(..), parseExpWithMode)

import Flame.IFC

myDefaultParseMode2 :: ParseMode
myDefaultParseMode2 = ParseMode
  {parseFilename = []
  ,baseLanguage = Haskell2010
  ,extensions = map EnableExtension myDefaultExtensions2
  ,ignoreLinePragmas = False
  ,ignoreLanguagePragmas = False
  ,fixities = Nothing
  ,ignoreFunctionArity = False
  }

myDefaultExtensions2 :: [KnownExtension]
myDefaultExtensions2 = [Ext.PostfixOperators
                      ,Ext.QuasiQuotes
                      ,Ext.UnicodeSyntax
                      ,Ext.PatternSignatures
                      ,Ext.MagicHash
                      ,Ext.ForeignFunctionInterface
                      ,Ext.TemplateHaskell
                      ,Ext.RankNTypes
                      ,Ext.MultiParamTypeClasses
                      ,Ext.RecursiveDo
                      ,Ext.TypeApplications]
  
myParseExp :: String -> Either String Exp
myParseExp = either Left (Right . toExp) . myParseHsExp

myParseHsExp :: String -> Either String (Hs.Exp Hs.SrcSpanInfo)
myParseHsExp = parseResultToEither . parseExpWithMode myDefaultParseMode2

flame :: QuasiQuoter
flame = QuasiQuoter { quoteExp = \str -> case myParseExp str of
        Right (DoE Nothing ss) -> return (go ss)
        Right _ -> fail "Expecting do notation"
        Left err -> fail (show err)
        , quotePat = undefined
        , quoteType = \str -> case parseType str of
        Right t -> return t
        Left err -> fail (show err)
        , quoteDec = undefined
        } where
    go [NoBindS e] = e
    go (BindS p e : ss) = VarE 'use `AppE` e `AppE` LamE [p] (go ss)
    go (NoBindS e : ss) = VarE 'apply `AppE` e `AppE` LamE [WildP] (go ss)
    go (LetS ds : ss) = LetE ds (go ss)
