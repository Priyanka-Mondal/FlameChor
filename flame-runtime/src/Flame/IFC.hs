{-# LANGUAGE GADTs, PolyKinds, KindSignatures, MultiParamTypeClasses,
    DataKinds, RankNTypes, FlexibleInstances, FlexibleContexts, TypeFamilies #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE UndecidableInstances #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# OPTIONS_GHC -fplugin Flame.Solver -fobject-code #-}

module Flame.IFC
       ( Labeled(label, unlabel,relabel), Lbl
       , IFC(lift, apply, bind, runIFC, protect, use, reprotect, ffmap, fjoin)
       , FLA(..), FLAC, FLACT, NMIF(..), NM, NMT, runFLAC, runNM 
       , (:≽), (:=>=), (:⊑), (:<:)          -- Delegation types
       , Def(..), (≽), (=>=), (⊑), (<:)     -- Delegation constructors
       )
  where

import Flame.TCB.IFC
import Flame.TCB.Assume
