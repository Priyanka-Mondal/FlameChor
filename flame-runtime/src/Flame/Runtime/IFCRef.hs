{-# LANGUAGE DataKinds #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE RankNTypes #-}
-- {-# OPTIONS_GHC -fplugin Flame.Solver #-}

module Flame.Runtime.IFCRef
where

import Flame.Principals
import Flame.TCB.IFC 
import Data.IORef

data IFCRef (l::KPrin) a = IFCRef { unsafeUnwrap :: IORef a}

newIFCRef :: forall l pc m n a. (IFC m IO n, pc ⊑ l) => a -> m IO n pc pc (IFCRef l a)
newIFCRef a = unsafeProtect $ do 
                  r <- newIORef a
                  return . label $ IFCRef r

writeIFCRef :: (IFC m IO n, pc ⊑ l) => IFCRef l a -> a -> m IO n pc pc' ()
writeIFCRef ref a = unsafeProtect $ do 
                      writeIORef (unsafeUnwrap ref) a;
                      return . label $ ()

readIFCRef :: IFC m IO n => IFCRef l a -> m IO n pc (pc ⊔ l) a
readIFCRef ref = unsafeProtect $ do 
                   r <- readIORef (unsafeUnwrap ref)
                   return . label $ r
--modifyIORef :: IORef a -> (a -> a) -> IO ()
