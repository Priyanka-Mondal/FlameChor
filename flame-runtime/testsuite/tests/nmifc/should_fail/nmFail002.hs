{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE PostfixOperators #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE RankNTypes #-}
{-# OPTIONS_GHC -fplugin Flame.Solver #-}
module ShouldFail where

import Flame.Principals
import Flame.Runtime.Principals
import Flame.IFC 

{- #2 -}
nm_two :: (BFLA c m e n, q ⊑ Δ q) => SPrin p -> SPrin q
       -> c m e n (Δ q) (I p) (C p ∧ (I (p ∨ q))) a
       -> c m e n (Δ q) (I p) p a
nm_two p q v = iassume ((q*←) ≽ (p*←)) (reprotect_b v)
