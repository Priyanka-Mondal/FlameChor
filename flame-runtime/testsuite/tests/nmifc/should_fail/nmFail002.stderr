
nmFail002.hs:18:41:
     Could not deduce (C p ≽ C (C p ∧ I (p ∨ q)),
                        I (C p ∧ I (p ∨ q)) ≽ I p)
        arising from a use of ‘reprotect_b’
      from the context: (BFLA c m e n, q ⊑ Δ q)
        bound by the type signature for:
                   nm_two :: (BFLA c m e n, q ⊑ Δ q) =>
                             SPrin p
                             -> SPrin q
                             -> c m e n (Δ q) (I p) (C p ∧ I (p ∨ q)) a
                             -> c m e n (Δ q) (I p) p a
        at nmFail002.hs:(15,1)-(17,33)
      or from: I q ≽ I p
        bound by a type expected by the context:
                   (I q ≽ I p) => c m e n (Δ q) (I p) p a
        at nmFail002.hs:18:16-54
     In the second argument of ‘iassume’, namely ‘(reprotect_b v)’
      In the expression: iassume ((q *←) ≽ (p *←)) (reprotect_b v)
      In an equation for ‘nm_two’:
          nm_two p q v = iassume ((q *←) ≽ (p *←)) (reprotect_b v)
